### ISSUE TITLE
Ping fails on 1.4.1

### ISSUE THREAD

**Issue by clawster** [Textalk/websocket-php#108]
We have an application that uses ping to check that the remote endpoint is alive, but this stopped working with 1.4.1. 

I have tracked it down to how the response is handled in 1.4.1 

        // if we received a pong, wait for the next message
        if ($opcode === 'pong') {
            $this->logger->debug("Received 'pong'.");
            return [null, false];
        }

Tried this fix, and it seem to work here ^tm

        // if we received a pong
        if ($opcode === 'pong') {
            $this->logger->debug("Received 'pong'.");
            $this->last_opcode=$opcode;
            return [null, true];
        }

We have a ping method that looks like this...

    /**
     *
     * https://tools.ietf.org/html/rfc6455#section-5.5.2
     *
     * A ping response is not prioritized
     *
     * @param \WebSocket\Client $client
     * @return bool
     * @throws ConnectionException
     */
    static public function ping(\WebSocket\Client $client) : bool
    {

        // send and receive methods throws ConnectionException on connection errors.
        // Send ping OPCODE, don't care about the response payload, just the response opcode.
        // The send method lacks support for zero payload, so lets send a friendly string.

        $client->send("hi", self::OPCODE_STR_PING);
        $client->setTimeout(5);
        $client->receive(); // Response should be equal to what we sent.

        // Lets just look for the PONG OPCODE
        if ($client->getLastOpcode() !== self::OPCODE_STR_PONG) {
            throw new ConnectionException("Error: WebSocket Ping failed");
        }

        return true;
    }

Edited: Internal ping response handling makes sense, but in our use-case the pong must be returned to the caller.

**Comment by sirn-se** [comment]
Yes, I see the problem. 

We stopped returning control messages in `v1.4.1`. This was because they could break fragmented content messages. In short, the `$client->receive()` in your code could just as well pick up a continuous text or binary frame as a pong frame.
Besides breaking fragmentet content messages, your ping() method could throw exception even if the connection is all fine.

Unfortunately these issues can't be properly fixed without a significant re-write of the code. The patch you suggest would only reintroduce the problem we attempted to fix to begin with.

**Comment by clawster** [comment]
Thanks for your feedback. 

In our scenario we have had connection issues. Probably due to that the php-application does not talk directly to the remote application but through a AWS ALB -> nginx proxy -> rust application.

It seems that sometimes we got a connection but the rust application is not available.
So we want to verify (with a ping) that we are talking with the rust application directly after connecting.

Implement a custom "hello" endpoint does not feel right in our scenario, we do not want to add that extra complexity unless strictly necessary.

Would it be an option for you to add a $client->isAlive() method to make it possible to read the pong response? Maybe something like this?

```
diff --git a/lib/Base.php b/lib/Base.php
index 05058c5..d959348 100644
--- a/lib/Base.php
+++ b/lib/Base.php
@@ -19,6 +19,8 @@ class Base implements LoggerAwareInterface
     protected $options = [];
     protected $is_closing = false;
     protected $last_opcode = null;
+    protected $last_ping = -1;
+    protected $last_pong = -1;
     protected $close_status = null;
     protected $logger;
 
@@ -36,6 +38,16 @@ class Base implements LoggerAwareInterface
         return $this->last_opcode;
     }
 
+    public function getLastPing(): int
+    {
+        return $this->last_ping;
+    }
+
+    public function getLastPong(): int
+    {
+        return $this->last_pong;
+    }
+
     public function getCloseStatus(): ?int
     {
         return $this->close_status;
@@ -171,7 +183,7 @@ class Base implements LoggerAwareInterface
         return $payload;
     }
 
-    protected function receiveFragment(): array
+    private function receiveData(): array
     {
         // Just read the main fragment information first.
         $data = $this->read(2);
@@ -229,9 +241,30 @@ class Base implements LoggerAwareInterface
             }
         }
 
+        // Return data
+        return [
+            'opcode' => $opcode,
+            'payload' => $payload,
+            'payload_length' => $payload_length,
+            'final' => $final,
+        ];
+    }
+
+    protected function receiveFragment(): array
+    {
+
+        // Receive data
+        $data = $this->receiveData();
+
+        $opcode = $data['opcode'];
+        $payload = $data['payload'];
+        $payload_length = $data['payload_length'];
+        $final = $data['final'];
+
         // if we received a ping, send a pong and wait for the next message
         if ($opcode === 'ping') {
             $this->logger->debug("Received 'ping', sending 'pong'.");
+            $this->last_ping = time();
             $this->send($payload, 'pong', true);
             return [null, false];
         }
@@ -239,6 +272,7 @@ class Base implements LoggerAwareInterface
         // if we received a pong, wait for the next message
         if ($opcode === 'pong') {
             $this->logger->debug("Received 'pong'.");
+            $this->last_pong = time();
             return [null, false];
         }
 
@@ -277,6 +311,31 @@ class Base implements LoggerAwareInterface
         return [$payload, $final];
     }
 
+    /**
+     * Send ping to remote endpoint and expect pong
+     *
+     * @throws ConnectionException  When unable to get a pong response
+     * @return void
+     */
+    public function isAlive(): void
+    {
+
+        // Send ping
+        $this->send("", "ping");
+
+        // Receive pong
+        $data = $this->receiveData();
+
+        // if we received a pong, wait for the next message
+        if ($data['opcode'] === 'pong') {
+            $this->logger->debug("Received 'pong'.");
+            $this->last_pong = time();
+            return;
+        }
+
+        throw new ConnectionException("Did not receive pong from remote");
+    }
+
     /**
      * Tell the socket to close.
      *
```
ps. maybe using private class variables would be a bit prettier, but just wanted to get your thoughts on this.

**Comment by sirn-se** [comment]
Unfortunately the issue would remain with above code. Code can not rely on the next stream read to be of a specific type.

I have been working on a sustainable solution for this (and some other) issues, using a listener and internal buffering. But that's a major re-write and I can't say when it will be completed.

So I would recommend locking to `1.4.0` as of now, if your need for ping/pong operation is more important than avoiding broken content messages. Or rely on the timeout kicking in if the server don't respond - any valid response (not just pong) would indicate the server is alive.

Sorry for the inconvenience caused by the 1.4.1 fix.

**Comment by sirn-se** [comment]
@clawster Just released version `1.5` which should provide the features you need.

It adds a filter option that tells receive() method which types of messages to return. Default to content message, but by setting `'filter' => ['text', 'binary', 'ping']` it will return ping messages as well.
It also fixes the issue with fragmented messages getting broken.

**Comment by clawster** [comment]
Big thanks, will try it out!

//Claes
